### PR TITLE
Gamestates

### PR DESCRIPTION
**Pull request by wouterz**
Game now has states: Start (can contain a menu/lobby in the future), Play, Won and Lost

### CODE REVIEW

**Comment by dmalarev** [comment]
LGTM

File: src/main/java/nl/tudelft/pixelperfect/gamestates/PlayState.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+import com.jme3.scene.Spatial;
+import jmevr.app.VRApplication;
+import nl.tudelft.pixelperfect.Game;
+import nl.tudelft.pixelperfect.Spaceship;
+import nl.tudelft.pixelperfect.event.EventScheduler;
+import nl.tudelft.pixelperfect.gui.GameHeadsUpDisplay;
+
+/**
+ * Created by woute on 5/23/2016.
```
Comment (jessetilro) [inline src/main/java/nl/tudelft/pixelperfect/gamestates/PlayState.java]: Can you please refactor alle these class scoped javadoc comments to be consistent with the rest of the code? It should state in a few sentences what the purpose and responsibilities of the class are. Then after a white line there should be an author tag followed by your full name.

File: src/test/java/nl/tudelft/pixelperfect/gamestates/WonStateTest.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  /**
+   * Test that you stay in WonState if key is not pressed.
+   */
+  private void handleStateThis() {
+    when(mockGame.isStartKey()).thenReturn(false);
+    GameState newState = testState.handleState();
+    assertSame(newState.getClass(), WonState.class);
+  }
+
+}
```
Comment (dalderliesten) [inline src/test/java/nl/tudelft/pixelperfect/gamestates/WonStateTest.java]: Please don't forget to keep an empty line, it's in our code convention so far.

File: src/test/java/nl/tudelft/pixelperfect/gamestates/PlayStateTest.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+import org.mockito.Mockito;
+
+import static org.junit.Assert.assertSame;
+import static org.mockito.Matchers.anyFloat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * Created by woute on 5/24/2016.
```
Comment (dalderliesten) [inline src/test/java/nl/tudelft/pixelperfect/gamestates/PlayStateTest.java]: Please fix this javadoc.

File: src/test/java/nl/tudelft/pixelperfect/gamestates/StartStateTest.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import nl.tudelft.pixelperfect.Game;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.assertSame;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Created by woute on 5/24/2016.
```
Comment (dalderliesten) [inline src/test/java/nl/tudelft/pixelperfect/gamestates/StartStateTest.java]: Please fix this javadoc.

File: src/test/java/nl/tudelft/pixelperfect/gamestates/WonStateTest.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import nl.tudelft.pixelperfect.Game;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.assertSame;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Created by woute on 5/24/2016.
```
Comment (dalderliesten) [inline src/test/java/nl/tudelft/pixelperfect/gamestates/WonStateTest.java]: Please fix this javadoc.

File: src/test/java/nl/tudelft/pixelperfect/gamestates/LostStateTest.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import nl.tudelft.pixelperfect.Game;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.assertSame;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Created by woute on 5/24/2016.
```
Comment (dalderliesten) [inline src/test/java/nl/tudelft/pixelperfect/gamestates/LostStateTest.java]: Please fix this javadoc.

File: src/test/java/nl/tudelft/pixelperfect/gamestates/GameStateTest.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package nl.tudelft.pixelperfect.gamestates;
+
+import org.junit.Test;
+
+/**
+ * Created by woute on 5/24/2016.
```
Comment (dalderliesten) [inline src/test/java/nl/tudelft/pixelperfect/gamestates/GameStateTest.java]: Please fix this javadoc.

File: src/main/java/nl/tudelft/pixelperfect/gamestates/WonState.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+import nl.tudelft.pixelperfect.Game;
+
+/**
+ * Created by woute on 5/23/2016.
+ */
+public class WonState extends GameState {
+
+  /**
+   * Constructor for WonState.
+   * @param game
```
Comment (dalderliesten) [inline src/main/java/nl/tudelft/pixelperfect/gamestates/WonState.java]: Please add a space between the javadoc description and the javadoc input/param/etc list.

File: src/main/java/nl/tudelft/pixelperfect/gamestates/StartState.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
+package nl.tudelft.pixelperfect.gamestates;
+
+
+import nl.tudelft.pixelperfect.Game;
+
+/**
+ * Created by woute on 5/23/2016.
```
Comment (dalderliesten) [inline src/main/java/nl/tudelft/pixelperfect/gamestates/StartState.java]: Please fix this javadoc.

File: src/main/java/nl/tudelft/pixelperfect/gamestates/PlayState.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    updateMovement(tpf);
+
+    scheduler.update(tpf);
+    spaceship.update(tpf);
+
+    // Update the in-game heads up display.
+    headsUpDisplay.updateHud();
+
+//    for (Event event : spaceship.getLog().getEvents()) {
+//      event.notification(game.getScene());
```
Comment (dalderliesten) [inline src/main/java/nl/tudelft/pixelperfect/gamestates/PlayState.java]: Why is this commented out?

File: src/main/java/nl/tudelft/pixelperfect/gamestates/LostState.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
+package nl.tudelft.pixelperfect.gamestates;
+
+import nl.tudelft.pixelperfect.Game;
+
+/**
+ * State for when the game is lost.
+ * Created by woute on 5/23/2016.
```
Comment (dalderliesten) [inline src/main/java/nl/tudelft/pixelperfect/gamestates/LostState.java]: Please fix this javadoc, although it's already better than the rest.

File: src/main/java/nl/tudelft/pixelperfect/gamestates/PlayState.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+public class PlayState extends GameState {
+
+  private Spatial observer;
+  private Spaceship spaceship;
+  private GameHeadsUpDisplay headsUpDisplay;
+  private EventScheduler scheduler;
+
+  /**
+   * Constructor for PlayState.
+   * @param game
```
Comment (dalderliesten) [inline src/main/java/nl/tudelft/pixelperfect/gamestates/PlayState.java]: Please watch the spacing here, see previous comment.

File: src/main/java/nl/tudelft/pixelperfect/Game.java [hunk 13]
```diff
@@ -170,36 +185,107 @@ public Spaceship getSpaceship() {
...
+   */
+  public boolean isRotateRight() {
+    return rotateRight;
+  }
+
+  /**
+   * Getter for startKey.
+   * @return startKey
+   */
+  public boolean isStartKey() {
```
Comment (jessetilro) [inline src/main/java/nl/tudelft/pixelperfect/Game.java]: Method names like `isStartKey` and `isRotateLeft` are quite vague. The Game "is" not a start key, it has its start key pressed for example. So a better name in my opinion would be along the lines of `isStarted` or `isReady` or `hasBeenStarted`. Likewise I suggest to rename the other boolean getter methods.

**Comment by dalderliesten** [comment]
If you can fix the javadoc issues I pointed out, I approve.  The functionality is good!

File: src/main/java/nl/tudelft/pixelperfect/Game.java (hunk 13 above)
Comment (wouterz) [inline src/main/java/nl/tudelft/pixelperfect/Game.java]: Both are booleans linked to keypresses, so basically "is the start key pressed", "is rotate left key pressed"

**Comment by FlorisXll** [comment]
Whenever I hold the "P" button after the game was lost I can continue to play the game.

**Pull request merged**